pkg/model/multiversion/delta.go

```diff
@@ -0,0 +1,289 @@
+	FieldChangeTypeNone FieldChangeType = "none"
+	// FieldChangeTypeAdded is used when a new field is introduced in a CRD.
+	FieldChangeTypeAdded FieldChangeType = "added"
+	// FieldChangeTypeRemoved is used a when a field is removed from a CRD.
+	FieldChangeTypeRemoved FieldChangeType = "removed"
+	// FieldChangeTypeRenamed is used when a field is renamed.
+	FieldChangeTypeRenamed FieldChangeType = "renamed"
+	// FieldChangeTypeShapeChanged is used when a field shape has changed.
+	FieldChangeTypeShapeChanged FieldChangeType = "shape-changed"
+	// FieldChangeTypeShapeChangedFromStringToSecret is used when a field change to
+	// a k8s secret type.
+	FieldChangeTypeShapeChangedFromStringToSecret FieldChangeType = "shape-changed-from-string-to-secret"
+	// FieldChangeTypeShapeChangedFromSecretToString is used when a field changed from
+	// a k8s secret to a Go string.
+	FieldChangeTypeShapeChangedFromSecretToString FieldChangeType = "shape-changed-from-secret-to-string"
+)
+
+// FieldDelta represent the delta between the same field in two different
+// CRD versions. If a field is removed in the Destination version the Destination value will
+// be nil. If a field is new in the Destination version, the Source value will be nil.
+type FieldDelta struct {
+	ChangeType FieldChangeType
+	// Field from the destination CRD
+	Destination *ackmodel.Field
+	// Field from the source CRD
+	Source *ackmodel.Field
+}
+
+// CRDDelta stores the spec and status deltas for a custom resource.
+type CRDDelta struct {
+	SpecDeltas   []FieldDelta
+	StatusDeltas []FieldDelta
+}
+
+// ComputeCRDFieldsDeltas compares two ackmodel.CRD instances and returns the
+// spec and status fields deltas. srcCRD is the CRD of the spoke (source) version
+// and dstCRD is the CRD of the hub (destination) version.
+func ComputeCRDFieldsDeltas(srcCRD, dstCRD *ackmodel.CRD) (*CRDDelta, error) {
+	dstRenames, err := dstCRD.GetAllRenames()
+	if err != nil {
+		return nil, fmt.Errorf("cannot get resource field renames: %s", err)
+	}
+
+	srcRenames, err := srcCRD.GetAllRenames()
+	if err != nil {
+		return nil, fmt.Errorf("cannot get resource field renames: %s", err)
+	}
+
+	renames, err := ComputeRenamesDelta(srcRenames, dstRenames)
+	if err != nil {
+		return nil, fmt.Errorf("cannot compute the field renames delta: %v", err)
+	}
+
+	specDeltas, err := ComputeFieldsDelta(srcCRD.SpecFields, dstCRD.SpecFields, renames)
+	if err != nil {
+		return nil, fmt.Errorf("cannot compute spec fields deltas: %s", err)
+	}
+
+	statusDeltas, err := ComputeFieldsDelta(srcCRD.StatusFields, dstCRD.StatusFields, renames)
+	if err != nil {
+		return nil, fmt.Errorf("cannot compute status fields deltas: %s", err)
+	}
+
+	return &CRDDelta{
+		specDeltas,
+		statusDeltas,
+	}, nil
+}
+
+// fieldChangedToSecret returns true if field changed from string to secret.
+func fieldChangedToSecret(src, dst *ackmodel.Field) bool {
+	return (src.FieldConfig == nil ||
```

Comment:
`src.FieldConfig == nil` implies this is a new field? Not that it changed from `string` to `Secret`? Is this the intended effect?

Reply:
>src.FieldConfig == nil implies this is a new field? 

Not really, a field can exist and have no configuration in `generator.yaml`.

>  Not that it changed from string to Secret?

This function return true if a field changed from `string` to `Secret`. To find these fields we should verify that either the `src` Field have a `is_secret: false` or don't have a configuration in the `generator.yaml` (which is equivalent to `is_secret: false`), and that `dst` have a `is_secret: true`